### PR TITLE
[WIP] seq2seq example

### PR DESCRIPTION
**Pull request by mgoldey**
This PR presents an example seq2seq use case and bug fixes necessary for this to execute with reasonable accuracy. 

The utils_seq2seq.py file defines the data format for training data, and the run_seq2seq.py file takes training, development, and test data and produces a model. The README.md discusses how to execute this toy problem. The specific toy problem in use here is formatting a date string to the American style, which is a trivial example. On my local setup using GPUs, this example executes within 5 minutes. Production models should include more learnings.

I welcome feedback about how to strengthen performance here and the best route to increase testing.

This relies on a few bug fixes which have been incorporated in this branch
- Without a fix for #3038, PreTrainedEncoderDecoder won't instantiate at all. 
- Without a fix for #2435, BERT models fail completely on this use case as the BERT decoder isn't instantiated correctly without CrossAttention in that case.
- ~I strongly suspect that the input to the decoder in the PreTrainedEncoderDecoder class is incorrect as present in the code base, and commit https://github.com/huggingface/transformers/commit/9fcf73afbcfa18918234592039da7bd409820431 has a proposed fix. It doesn't make sense to have the expected token ids as input to the decoder when the decoder needs to learn how to decode from the embeddings.~ Incomplete understanding - will fix

### CODE REVIEW

**Comment by mgoldey** [comment]
As a non-blocking question, I do note that a lot of the examples use argparse to parse comparatively long lists of arguments. I've maintained the extant style in this PR to avoid causing noise and confusion Would it be acceptable if I broke with this style to use a JSON file to store all the arguments for an experiment?

**Comment by patrickvonplaten** [comment]
Hi @mgoldey, sorry for only responding now. Thanks a lot for adding a seq2seq example :-) I will take a look early next week and maybe we can have a quick chat how to merge this PR and https://github.com/huggingface/transformers/pull/3383.

**Comment by mgoldey** [comment]
That sounds good. I'm still tweaking things on my end for accuracy and improved logic as I get more familiar with the code base here. I'll see if I can rebase of #3383 by then, depending on my other workload. Feel free to reach out via google hangouts if you're comfortable. What would help get this over to #3383 is advice on correctly constructing the same `prepare_inputs_for_generation` in modeling_gpt2.py.

File: examples/seq2seq/data_generator.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+CONSTANTS["months"] = {
+    1: lambda: "january" if random.random() > 0.7 else "jan",
+    2: lambda: "february" if random.random() > 0.7 else "feb",
+    3: lambda: "march" if random.random() > 0.7 else "mar",
+    4: lambda: "april" if random.random() > 0.7 else "apr",
+    5: lambda: "may",
+    6: lambda: "june" if random.random() > 0.7 else "jun",
+    7: lambda: "july" if random.random() > 0.7 else "jul",
+    8: lambda: "august" if random.random() > 0.9 else ("auggie" if random.random() > 0.5 else "aug"),
```
Comment (sshleifer) [inline examples/seq2seq/data_generator.py]: Are these randomly inconsistent constants intentional?

File: examples/seq2seq/utils_seq2seq.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    """
+    Loads a data file into a list of `InputBatch`s
+    """
+    #  `cls_token_segment_id` define the segment id associated to the CLS token (0 for BERT, 2 for XLNet)
+    cls_token_segment_id = 2 if model_type in ["xlnet"] else 0
+    pad_token_segment_id = 4 if model_type in ["xlnet"] else 0
+    features = []
+
+    cls_token, sep_token, pad_token = cls_token = (
+        tokenizer.cls_token,
```
Comment (dbaxter240) [inline examples/seq2seq/utils_seq2seq.py:99]: Hi, I've been using your example to guide my own setup of a seq2seq model (thank you!) Is this line in your pull request a typo? Think it's just supposed to be cls_token, sep_token, pad_token = ( ...

File: examples/seq2seq/utils_seq2seq.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        # defines the location of the CLS token:
+        # - False (Default, BERT/XLM pattern): [CLS] + A + [SEP] + B + [SEP]
+        # - True (XLNet/GPT pattern): A + [SEP] + B + [SEP] + [CLS]
+        # xlnet/gpt is special
+        if model_type in ["xlnet"]:
+            tokens += [cls_token]
+            formatted_tokens += [pad_token_label_id]
+            segment_ids += [cls_token_segment_id]
+        # gpt2 has no cls_token
+        elif model_type not in ["gpt2"]:
```
Comment (dbaxter240) [inline examples/seq2seq/utils_seq2seq.py:157]: I believe this is supposed to be 

elif model_type in ['gpt2']

 instead of "not in"

**Comment by patrickvonplaten** [comment]
Sorry, to answer only now! I'll will soon add a Encoder-Decoder google colab that shows how to use seq2seq

**Comment by mgoldey** [comment]
Thanks - fine to close. We've moved forward without using seq2seq due to poor overall accuracy with the scale of data in place.
